safenode/src/node/replication.rs

```diff
@@ -0,0 +1,155 @@
+    protocol::messages::{
+        DataRequest, DataResponse, Event, Query, QueryResponse, RegisterQuery, Request, Response,
+    },
+};
+use libp2p::PeerId;
+use std::{collections::HashSet, time::Duration};
+use tokio::spawn;
+
+pub(crate) async fn send_out_data_addrs(
+    network: Network,
+    chunk_storage: ChunkStorage,
+    _register_storage: RegisterStorage,
+) {
+    tokio::time::sleep(Duration::from_secs(10)).await;
+    let mut our_closest_peers = HashSet::new();
+    if let Ok(closest) = network
+        .get_closest_local_peers(network.peer_id.to_bytes(), false)
+        .await
+    {
+        our_closest_peers = closest.into_iter().collect();
+    }
+    loop {
+        let new_peers = check_for_group_change(&network, &mut our_closest_peers).await;
+        info!("Churn detected, the new_peers in our closest_group is: {new_peers:?}");
+        // send out our data address to them
+        // todo: should do the same for reg storage
+        let data_addresses: Vec<DataAddress> = chunk_storage
+            .addrs()
+            .into_iter()
+            .map(DataAddress::Chunk)
+            .collect();
+        for peer in new_peers {
+            let network_clone = network.clone();
+            let data_addresses = data_addresses.clone();
+            // spawn task to send event to each peer
+            let _handle = spawn(async move {
+                let _ = network_clone
+                    .send_request(
+                        Request::Event(Event::ReplicateData(data_addresses.clone())),
+                        peer,
+                    )
+                    .await;
+            });
+        }
+        tokio::time::sleep(Duration::from_secs(20)).await;
+    }
+}
+
+pub(crate) async fn ask_peers_for_data(
+    network: Network,
+    chunk_storage: ChunkStorage,
+    register_storage: RegisterStorage,
+    data_addresses: Vec<DataAddress>,
+) {
+    // for each addr, spawn a task to request for the data corresponding to the addr from its
+    // closest peers and store them.
+    for addr in data_addresses {
+        let req = match addr {
+            DataAddress::Chunk(addr) => {
+                if chunk_storage.get(&addr).await.is_ok() {
+                    continue;
+                }
+                Request::Data(DataRequest::Query(Query::GetChunk(addr)))
+            }
+            DataAddress::Register(addr) => {
+                if register_storage.addrs().await.contains(&addr) {
+                    continue;
+                }
+                Request::Data(DataRequest::Query(Query::Register(RegisterQuery::Get(
+                    addr,
+                ))))
+            }
+            DataAddress::Spend(_) => todo!(),
+        };
+        let network = network.clone();
+        let chunk_storage = chunk_storage.clone();
+        let _register_storage = register_storage.clone();
+        let _handle = spawn(async move {
+            if let Ok(responses) = network.node_send_to_closest(&req).await {
```

Comment:
this will send the req to close_group_peers, which is for sure a waste of resources and could trigger an avalanch that slow down the network a lot.
same we can discuss it further in the slack

Reply:
Good observations.